### ISSUE TITLE
Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory

### ISSUE THREAD

**Issue by ghost** [0x1Avram/js-deobfuscator#15]
When I deobfuscate the js file which is obfuscated by [obfuscator.io](https://obfuscator.io/), the following error occurs.

```
<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xb02960 node::Abort() [node]
 2: 0xa18149 node::FatalError(char const*, char const*) [node]
 3: 0xcdd22e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0xcdd5a7 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0xe94c15  [node]
 6: 0xe956f6  [node]
 7: 0xea3c1e  [node]
 8: 0xea4660 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 9: 0xea7555 v8::internal::Heap::HandleGCRequest() [node]
10: 0xe34bf7 v8::internal::StackGuard::HandleInterrupts() [node]
11: 0x11e0fd5 v8::internal::Runtime_StackGuard(int, unsigned long*, v8::internal::Isolate*) [node]
12: 0x15d54f9  [node]
zsh: IOT instruction  node index.js -i obfuscated.txt -o output.txt -s stringarray.txt -w  -n  -r
```

The command I use is:
> node index.js -i obfuscated.txt -o output.txt -s stringarray.txt -w stringarraywrappers.txt -n _0x1037 -r stringarrayrotate.txt

obfuscated.txt: https://pastebin.com/R0ZseBHB
stringarray.txt: https://pastebin.com/3CFk1s1v
stringarraywrappers.txt: https://pastebin.com/DcnV2QwZ
stringarrayrotate.txt: https://pastebin.com/pZ0PJ7xS

**Comment by 0x1Avram** [comment]
Hello, 
obfuscator.io adds some checks so that if the code is tampered with it crashes. Here it crashes when adding the code corresponding to the string rotate function to the execution context. Did you beautify/format the code? If so, do you have the original version?

As a side node, if you use the latest (1.0.5) version, the stringarray specific code(stringarray.txt, stringarraywrappers.txt, stringarrayrotate.txt) are recognized automatically, there is no need to manually pass them as input.

**Comment by ghost** [comment]
> As a side node, if you use the latest (1.0.5) version, the stringarray specific code(stringarray.txt, stringarraywrappers.txt, stringarrayrotate.txt) are recognized automatically, there is no need to manually pass them as input.

I use the latest version to deobfuscate this, it still crashes:
```
TypeError [ERR_INVALID_ARG_TYPE]: The "data" argument must be of type string or an instance of Buffer, TypedArray, or DataView. Received null
    at Object.writeFileSync (node:fs:2163:5)
    at Object.writeTextFile (/snapshot/js-deobfuscator/useful.js)
    at main (/snapshot/js-deobfuscator/index.js)
    at Object.<anonymous> (/snapshot/js-deobfuscator/index.js)
    at Module._compile (pkg/prelude/bootstrap.js:1930:22)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.runMain (pkg/prelude/bootstrap.js:1983:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'ERR_INVALID_ARG_TYPE'
}
```
I beautified the original code and the original version is here:
https://pastebin.com/3zNAaPCt

**Comment by 0x1Avram** [comment]
There was a problem while simplifying some numeric expressions that were getting evaluated as NaN.
Fixed it and created a new release - [1.0.6](https://github.com/0x1Avram/js-deobfuscator/releases/tag/1.0.6) The deobfuscated version looks like some sort of CTF challenge.
Another thing to note is that the deobfuscation only works on the orginal file and not the beautified one because of the anti tampering code I mentioned in the comment above.
You can run it like this:
`node index.js -i obfuscated.txt -o output.txt`
Please check it and let me know if you encounter other issues.

**Comment by ghost** [comment]
> The deobfuscated version looks like some sort of CTF challenge.

Yep, it is a simple CTF challenge and I have solved it yesterday using other techniques. I just check the new release(1.0.6), it works well and successfully recover the js file's original semantics. Anyway, your work works :)
